packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol

```diff
@@ -0,0 +1,249 @@
+                
+                // Load `amount`
+                let amount := calldataload(100)
+        
+                // Calculate number of bytes in `amounts` contents
+                let amountsByteLen := mul(amountsLen, 32)
+
+                // Loop through `amounts` and `nestedAssetData`, calling `transferFrom` for each respective element
+                for {let i := 0} lt(i, amountsByteLen) {i := add(i, 32)} {
+
+                    // Calculate the total amount
+                    let amountsElement := calldataload(add(amountsContentsStart, i))
+                    let totalAmount := mul(amountsElement, amount)
+
+                    // Revert if multiplication resulted in an overflow
+                    if iszero(eq(div(totalAmount, amount), amountsElement)) {
+                        // Revert with `Error("UINT256_OVERFLOW")`
+                        mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+                        mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                        mstore(64, 0x0000001055494e543235365f4f564552464c4f57000000000000000000000000)
+                        mstore(96, 0)
+                        revert(0, 100)
+                    }
+
+                    // Write `totalAmount` to memory
+                    mstore(100, totalAmount)
+
+                    // Load offset to `nestedAssetData[i]`
+                    let nestedAssetDataElementOffset := calldataload(add(nestedAssetDataContentsStart, i))
+
+                    // Load length of `nestedAssetData[i]`
+                    let nestedAssetDataElementContentsStart := add(nestedAssetDataElementOffset, 264)
+                    let nestedAssetDataElementLen := calldataload(sub(nestedAssetDataElementContentsStart, 32))
+
+                    // Revert if the `nestedAssetData` does not contain a 4 byte `assetProxyId`
+                    if lt(nestedAssetDataElementLen, 4) {
+                        // Revert with `Error("LENGTH_GREATER_THAN_3_REQUIRED")`
+                        mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+                        mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                        mstore(64, 0x0000001e4c454e4754485f475245415445525f5448414e5f335f524551554952)
+                        mstore(96, 0x4544000000000000000000000000000000000000000000000000000000000000)
+                        revert(0, 100)
+                    }
+
+                    // Load AssetProxy id
+                    let assetProxyId := and(calldataload(nestedAssetDataElementContentsStart), 0xffffffff00000000000000000000000000000000000000000000000000000000)
+
+                    // To lookup a value in a mapping, we load from the storage location keccak256(k, p),
+                    // where k is the key left padded to 32 bytes and p is the storage slot
+                    mstore(132, assetProxyId)
```

Comment:
132 is where `nestedAssetDataElementLen` should be stored for our outgoing calldata. `assetProxyId` would starting from 164.

Edit: doesn't matter, we overwrite anyway in the big calldatacopy. It does seem worth re-using these 4 bytes.